ipn/ipnlocal/serve.go

```diff
@@ -271,154 +276,63 @@ func (b *LocalBackend) DeleteForegroundSession(sessionID string) error {
-	setHandler(sc, req)
-	if err := b.SetServeConfig(sc); err != nil {
-		return fmt.Errorf("errro setting serve config: %w", err)
+	// we can't have multiple TCPs under a single foreground session
+	var port uint16
+	for key := range fg.TCP {
+		port = key
 	}
-	// Defer turning off Funnel once stream ends.
-	defer func() {
-		sc := b.ServeConfig().AsStruct()
-		deleteHandler(sc, req, port)
-		err = errors.Join(err, b.SetServeConfig(sc))
-	}()
-
-	var writeErrs []error
-	writeToStream := func(log ipn.FunnelRequestLog) {
-		jsonLog, err := json.Marshal(log)
-		if err != nil {
-			writeErrs = append(writeErrs, err)
-			return
-		}
-		if _, err := fmt.Fprintf(w, "%s\n", jsonLog); err != nil {
-			writeErrs = append(writeErrs, err)
-			return
+	var hasFunnel bool
+	for key, b := range sc.AllowFunnel {
+		givenPort, _ := key.Port()
+		if givenPort == port {
+			hasFunnel = b
+			hp = key
+			break
 		}
-		f.Flush()
-	}
-
-	// Hook up connections stream.
-	b.mu.Lock()
-	mak.NonNilMapForJSON(&b.serveStreamers)
-	if b.serveStreamers[port] == nil {
-		b.serveStreamers[port] = make(map[uint32]func(ipn.FunnelRequestLog))
 	}
-	id := uuid.New().ID()
-	b.serveStreamers[port][id] = writeToStream
-	b.mu.Unlock()
-
-	// Clean up streamer when done.
-	defer func() {
-		b.mu.Lock()
-		delete(b.serveStreamers[port], id)
-		b.mu.Unlock()
-	}()
-
-	select {
-	case <-ctx.Done():
-		// Triggered by foreground `tailscale funnel` process
-		// (the streamer) getting closed, or by turning off Tailscale.
+	if !hasFunnel {
+		return "", false
 	}
-
-	return errors.Join(writeErrs...)
-}
-
-func setHandler(sc *ipn.ServeConfig, req ipn.ServeStreamRequest) {
-	if sc.TCP == nil {
-		sc.TCP = make(map[uint16]*ipn.TCPPortHandler)
+	if _, ok := sc.TCP[port]; ok {
+		return "", false
 	}
-	if _, ok := sc.TCP[443]; !ok {
-		sc.TCP[443] = &ipn.TCPPortHandler{
-			HTTPS: true,
+	for key, givenFg := range sc.Foreground {
+		if key == sessionID {
+			continue
 		}
-	}
-	if sc.Web == nil {
-		sc.Web = make(map[ipn.HostPort]*ipn.WebServerConfig)
-	}
-	wsc, ok := sc.Web[req.HostPort]
-	if !ok {
-		wsc = &ipn.WebServerConfig{}
-		sc.Web[req.HostPort] = wsc
-	}
-	if wsc.Handlers == nil {
-		wsc.Handlers = make(map[string]*ipn.HTTPHandler)
-	}
-	wsc.Handlers[req.MountPoint] = &ipn.HTTPHandler{
-		Proxy: req.Source,
-	}
-	if req.Funnel {
-		if sc.AllowFunnel == nil {
-			sc.AllowFunnel = make(map[ipn.HostPort]bool)
+		if _, ok := givenFg.TCP[port]; ok {
+			return "", false
 		}
-		sc.AllowFunnel[req.HostPort] = true
 	}
+	return hp, true
 }
 
-func deleteHandler(sc *ipn.ServeConfig, req ipn.ServeStreamRequest, port uint16) {
-	delete(sc.AllowFunnel, req.HostPort)
-	if sc.TCP != nil {
-		delete(sc.TCP, port)
-	}
-	if sc.Web == nil {
-		return
-	}
-	if sc.Web[req.HostPort] == nil {
-		return
-	}
-	wsc, ok := sc.Web[req.HostPort]
-	if !ok {
-		return
-	}
-	if wsc.Handlers == nil {
-		return
+func (b *LocalBackend) maybeLogServeConnection(destPort uint16, srcAddr netip.AddrPort, sessionID string) {
```

Comment:
document this? can sessionID be empty?

is this only called if the sessionID is alive? Is the session necessarily subscribing to access log traffic?

Reply:
Adding the following docs: 

```
// maybeLogServeConnection creates a RequestAccessLog and sends it to any watchers
// who are subscribed to ipn.NotifyServeRequest. SessionID, when not empty, indicates
// that this log is intended for a specific foreground serve session, otherwise it may be
// a background serve whose logs are being followed.
```

That said, I'm not sure how WatchIPNBus will work for when you want to stream the request logs of a background session?

Reply:
> how WatchIPNBus will work for when you want to stream the request logs of a background session?

In what way? You can not filter the access log stream if the CLI client wants all of them.

Reply:
We're punting on streaming request logs, but we should still chat about how to stream logs for both foreground and background sessions.